cudax/include/cuda/experimental/__memory_resource/cuda_async_memory_resource.h

```diff
@@ -0,0 +1,284 @@
+  {
+    return __pool_ != __rhs.__pool_;
+  }
+#    endif // _CCCL_STD_VER <= 2017
+
+  /**
+   * @brief Equality comparison between a cuda_memory_resource and another resource
+   * @param __lhs The cuda_memory_resource
+   * @param __rhs The resource to compare to
+   * @return If the underlying types are equality comparable, returns the result of equality comparison of both
+   * resources. Otherwise, returns false.
+   */
+  template <class _Resource>
+  _CCCL_NODISCARD_FRIEND auto operator==(cuda_async_memory_resource const& __lhs, _Resource const& __rhs) noexcept
+    _LIBCUDACXX_TRAILING_REQUIRES(bool)(_CUDA_VMR::__different_resource<cuda_async_memory_resource, _Resource>)
+  {
+    return _CUDA_VMR::resource_ref<>{const_cast<cuda_async_memory_resource&>(__lhs)}
+        == _CUDA_VMR::resource_ref<>{const_cast<_Resource&>(__rhs)};
+  }
+#    if _CCCL_STD_VER <= 2017
+  /**
+   * @copydoc cuda_async_memory_resource::operator==<_Resource>(cuda_memory_resource const&, _Resource const&)
+   */
+  template <class _Resource>
+  _CCCL_NODISCARD_FRIEND auto operator==(_Resource const& __rhs, cuda_async_memory_resource const& __lhs) noexcept
+    _LIBCUDACXX_TRAILING_REQUIRES(bool)(_CUDA_VMR::__different_resource<cuda_async_memory_resource, _Resource>)
+  {
+    return _CUDA_VMR::resource_ref<>{const_cast<cuda_async_memory_resource&>(__lhs)}
+        == _CUDA_VMR::resource_ref<>{const_cast<_Resource&>(__rhs)};
+  }
+  /**
+   * @copydoc cuda_async_memory_resource::operator==<_Resource>(cuda_memory_resource const&, _Resource const&)
+   */
+  template <class _Resource>
+  _CCCL_NODISCARD_FRIEND auto operator!=(cuda_async_memory_resource const& __lhs, _Resource const& __rhs) noexcept
+    _LIBCUDACXX_TRAILING_REQUIRES(bool)(_CUDA_VMR::__different_resource<cuda_async_memory_resource, _Resource>)
+  {
+    return _CUDA_VMR::resource_ref<>{const_cast<cuda_async_memory_resource&>(__lhs)}
+        != _CUDA_VMR::resource_ref<>{const_cast<_Resource&>(__rhs)};
+  }
+  /**
+   * @copydoc cuda_async_memory_resource::operator==<_Resource>(cuda_memory_resource const&, _Resource const&)
+   */
+  template <class _Resource>
+  _CCCL_NODISCARD_FRIEND auto operator!=(_Resource const& __rhs, cuda_async_memory_resource const& __lhs) noexcept
+    _LIBCUDACXX_TRAILING_REQUIRES(bool)(_CUDA_VMR::__different_resource<cuda_async_memory_resource, _Resource>)
+  {
+    return _CUDA_VMR::resource_ref<>{const_cast<cuda_async_memory_resource&>(__lhs)}
+        != _CUDA_VMR::resource_ref<>{const_cast<_Resource&>(__rhs)};
+  }
+#    endif // _CCCL_STD_VER <= 2017
+
+  /**
+   * @brief Returns the underlying handle to the CUDA memory pool
+   */
+  _CCCL_NODISCARD constexpr cudaMemPool_t pool_handle() const noexcept
+  {
+    return __pool_;
+  }
+
+  /**
+   * @brief Enables the `device_accessible` property
+   */
+  friend constexpr void get_property(cuda_async_memory_resource const&, _CUDA_VMR::device_accessible) noexcept {}
```

Comment:
cudaMallocAsync can allocate host-accessible memory. For example I think you can use it for host pinned memory, and managed memory may be supported in the future.  Seems like the accessibility should depend on the memory pool for this MR.

Or are we limiting `cuda_async_mr` to only device-accessible memory? If it can be constructed from an existing `cudaMempool_t`, I don't see how we can.

Reply:
Urgh, I am not sure whether we can sniff that out statically

Reply:
I believe the only way we can do this is by adding a template parameter to the resource that defaults to device only and can be optionally provided by the user. 

We could then check in the constructor, whether the provided memory pool actually supports host allocations via `if 
constexpr`